### DISCUSSION TITLE
Can interactive transactions be used to emulate sub queries?

### DISCUSSION THREAD

**Discussion by blackhaj** [prisma/prisma#17404] [Q&A]
Hey folks,

I am trying to understand the pros/cons of using interactive transactions to emulate sub queries in Prisma.

## What I would like to achieve:

Here is a sample schema:
```prisma
model FirstQuestionAnswer {
    response Boolean
    respondentId String
}
model SecondQuestionAnswer {
    response Boolean
    respondentId String
}
```
I would like to do the following:
1. get all the `SecondQuestionAnswer`’s of all the respondents who answered true to the `FirstQuestionAnswer`
2. to do this in a way that means I don’t have to hold in memory all of the respondentIds (as it could be a lot of ids) hence why I would use a subquery normally
3. allow for pagination of the results as I will have millions of answers

Using two separate queries I would do the following:

```js
const firstQuestionResults = await prisma.firstQuestionAnswer.findMany({
    where: {
        response: true
    },
})
const desiredResults = await prisma.secondQuestionAnswer.findMany({
    where: {
        responderId: {
            in: convertToArrayOfRespondentIds(firstQuestionResults),
        }
    },
    // cursor based pagination
    take: 100,
    cursor: cursorFromPreviousCall,
})
```

I would like to do this in a single query as:
1. the number of respondentIds might be quite large so I don't want to have to hold them in memory
2. I would like to only make one call to the DB
3. I assume that postgres has built in caching/performance enhancements that would make it more efficient to leave postgres to handle it (particularly given I want to use pagination)

## My proposal with transactions:

```js
await this.prisma.$transaction(async (tx) => {
      const firstQuestionResults = await tx.firstQuestionAnswer.findMany({
        where: {
          response: true,
        },
      });
      const desiredResults = await tx.secondQuestionAnswer.findMany({
        where: {
          responderId: {
            in: convertToArrayOfRespondentIds(firstQuestionResults),
          },
        },
        // cursor based pagination
        take: 100,
        cursor: cursorFromPreviousCall,
      });
    });
```

This helps with points 1 and 2 above (no holding ids in memory and only 1 db call) I believe. Are there any implications for point 3? Is there a better way to achieve what I want?

Any help and input would be greatly appreciated. Thanks for the amazing tool.

**Comment by blackhaj** [comment]
I have been testing this out and it looks like this still results in two queries. 

So I think the answer to my question "can interactive transactions be user to emulate sub queries" is no

**Reply by nurul3101** [reply]
Hey @blackhaj 👋 
The queries invoked in the callback function of interactive transactions are wrapped in a single transaction, but the number of queries would be 2 in this case.
We have a feature request for Subqueries here:
- https://github.com/prisma/prisma/issues/6506

If you can add a 👍 to the feature request, it would help our product team prioritise it.

**Reply by blackhaj** [reply]
Done, thanks @nurul3101

**Reply by nurul3101** [reply]
No worries! 😄

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
